src/cpp/rtps/writer/ReaderProxy.cpp

```diff
@@ -397,78 +429,61 @@ bool ReaderProxy::requested_changes_set(
     return isSomeoneWasSetRequested;
 }
 
-bool ReaderProxy::process_initial_acknack()
+bool ReaderProxy::process_initial_acknack(
+        std::function<void(ChangeForReader_t& change)> func)
 {
     if (is_local_reader())
     {
-        return 0 != convert_status_on_all_changes(UNACKNOWLEDGED, UNSENT);
+        return 0 != convert_status_on_all_changes(UNACKNOWLEDGED, UNSENT, func);
     }
 
     return true;
 }
 
-bool ReaderProxy::set_change_to_status(
+void ReaderProxy::from_unsent_to_status(
         const SequenceNumber_t& seq_num,
         ChangeForReaderStatus_t status,
-        bool restart_nack_supression)
+        bool restart_nack_supression,
+        bool delivered)
 {
+    // This function must not be called by a best-effort reader.
+    // It will use acked_changes_set().
+    assert(is_reliable_);
+
     if (restart_nack_supression && is_remote_and_reliable())
     {
         assert(timers_enabled_.load());
         nack_supression_event_->restart_timer();
     }
 
-    if (seq_num <= changes_low_mark_)
-    {
-        return false;
-    }
-
+    // Called when delivering an UNSENT sample, the seq_number must exists in the ReaderProxy.
+    assert(seq_num > changes_low_mark_);
     ChangeIterator it = find_change(seq_num, true);
-    bool change_was_modified = false;
+    assert(changes_for_reader_.end() != it);
 
-    // If the status is UNDERWAY (change was right now sent) and the reader is besteffort,
-    // then the status has to be changed to ACKNOWLEDGED.
-    if (UNDERWAY == status)
+    if (it != changes_for_reader_.end())
     {
-        if (!is_reliable())
-        {
-            status = ACKNOWLEDGED;
-        }
-        else if (is_local_reader() || is_datasharing_reader())
+        assert(UNSENT == it->getStatus());
+
+        // Otherwise change status
+        if (it->getStatus() != status)
```

Comment:
If this method cannot be called to change the status from UNSENT to UNSENT, we may `assert(UNSENT != status);` instead of this condition

Reply:
Done